File: src/main/java/com/mtgames/firerpg/entities/Player.java
package com.mtgames.firerpg.entities;

import com.mtgames.firerpg.InputHandler;
import com.mtgames.firerpg.entities.particles.DashParticle;
import com.mtgames.firerpg.gfx.Screen;
import com.mtgames.firerpg.gfx.gui.Hud;
import com.mtgames.firerpg.level.Level;
import com.mtgames.firerpg.level.Script;

public class Player extends Mob {
	
	private final static String	NAME		= "Player";
	private final int			JUMPWAIT;
	private final int			JUMPSPEED;
	private final int			DASHSPEED;
	private final int			DASHWAIT;
	private final int			STAGGERLENGTH;
	
	private InputHandler		input;
	private Script				script;
	
	private int					xa			= 0;
	private int					xaDash		= 0;
	private int					ya			= 0;
	private int					dir;
	private int					modifier;
	private int					jumpWait	= 0;
	private int					dashWait	= 0;
	private int					dashTime	= 0;
	private int					staggerTime	= 0;
	public boolean				canJump		= false;
	public boolean				canDash		= true;
	public boolean				isStaggered	= false;
	public boolean				isDashing	= false;
	
	public Player(Level level, int x, int y, InputHandler input) {
		super(level, NAME, x, y);
		
		this.script = new Script("scripts/Player.js");
		script.doInit();
		
		JUMPWAIT = (int) script.get("JUMPWAIT");
		JUMPSPEED = (int) script.get("JUMPSPEED");
		DASHSPEED = (int) script.get("DASHSPEED");
		DASHWAIT = (int) script.get("DASHWAIT");
		STAGGERLENGTH = (int) script.get("STAGGERLENGTH");
		
		speed = (int) script.get("speed");
		xMin = (int) script.get("xMin");
		xMax = (int) script.get("xMax");
		yMin = (int) script.get("yMin");
		yMax = (int) script.get("yMax");
		
		this.level = level;
		this.input = input;
	}
	
	public void tick() {
		script.set("x", x);
		script.set("xa", xa);
		script.invoke("tick");
		xa = (int) script.get("xa");
		x = (int) script.get("x");
		
		if (input.reload.isPressed()) {
			script.load();
		}
		
		if (staggerTime == 0) {
			isStaggered = false;
		} else {
			isStaggered = true;
		}
		
		if (!isStaggered) {
			if (hasCollided(0, 1) && jumpWait > JUMPWAIT) {
				canJump = true;
				jumpWait = 0;
			}
			
			if (hasCollided(0, 1) && !canJump) {
				jumpWait += 1;
			}
			
			if (!hasCollided(0, 1)) {
				canJump = false;
			}
			
			if (input.space.isPressed() && canJump) {
				ya = -JUMPSPEED;
				canJump = false;
				animationFrame = 0;
			}
			
			if (input.up.isPressed() && canDash) {
				xaDash = DASHSPEED;
				canDash = false;
				dashWait = 0;
				animationFrame = 0;
			}
			
			if (input.left.isPressed()) {
				xa--;
			}
			
			if (input.right.isPressed()) {
				xa++;
			}
			
		} else {
			staggerTime -= 1;
			ya = 0;
		}
		
		if (xaDash != 0) {
			isDashing = true;
		} else {
			isDashing = false;
		}
		
		move(xa, ya);
		dash();
		ya = gravity(ya);
		
		if (isDashing) {
			int particleOffset;
			
			if (movingDir == 1) {
				particleOffset = -19;
			} else {
				particleOffset = 4;
			}
			
			for (int i = 0; i < 20; i++) {
				level.addParticle(new DashParticle(level, x, y, particleOffset));
			}
		}
	}
	
	public void render(Screen screen) {
		int xTile = 0;
		int yTile = 27;
		
		int xOffset = x;
		int yOffset = y;
		
		switch (movingDir) {
			case 0:
				dir = 0x01;
				modifier = 8;
				break;
			
			case 1:
				dir = 0x00;
				modifier = 0;
				break;
		}
		
		if (isJumping && !isDashing && !isStaggered) {
			xTile = 8;
		} else if (isDashing) {
			xTile = 12;
			animationFrame = 0;
		} else if (isStaggered) {
			xTile = 14;
			animationFrame = 0;
		}
		
		if (animationFrame == 1) {
			xTile += 2;
		} else if (animationFrame == 2) {
			xTile += 4;
		} else if (animationFrame == 3) {
			xTile += 6;
		}
		
		screen.render(xOffset - 8 + modifier, yOffset - 8, xTile + yTile * 32, dir);
		screen.render(xOffset - modifier, yOffset - 8, (xTile + 1) + yTile * 32, dir);
		screen.render(xOffset - 8 + modifier, yOffset, xTile + (yTile + 1) * 32, dir);
		screen.render(xOffset - modifier, yOffset, (xTile + 1) + (yTile + 1) * 32, dir);
		
		/* Draw overlay on character */
		/*
		 * screen.render(xOffset+modifier, yOffset, (xTile+2)+yTile*32, colour,
		 * dir);
		 */
		/*
		 * screen.render(xOffset+8-modifier, yOffset, (xTile+3)+yTile*32,
		 * colour, dir);
		 */
		
		double dashRatio = ((dashWait * 10d) / (DASHWAIT * 10d));
		Hud.renderDash(screen, dashRatio);
	}
	
	private void dash() {
		if (dashWait >= DASHWAIT || canDash == true) {
			dashWait = DASHWAIT;
		} else {
			dashWait++;
		}
		
		if (dashWait == DASHWAIT && !hasCollided(0, 1)) {
			canDash = true;
		} else {
			canDash = false;
		}
		
		if (xaDash == 0) {
			xa = 0;
			return;
		}
		
		if (hasCollided(-1, 0) || hasCollided(1, 0)) {
			staggerTime = STAGGERLENGTH;
			xaDash = 0;
		}
		
		ya = 0;
		gravityWait = 0;
		
		if (movingDir == 0) {
			xa = -xaDash;
		} else {
			xa = xaDash;
		}
		
		if (dashTime > 2) {
			xaDash--;
			dashTime = 0;
		}
		
		dashTime++;
	}
}

File: platformer/src/main/resources/python/entities/Player.py
from com.mtgames.platformer.scripting.interfaces import EntityInterface
from com.mtgames.platformer.gfx.lwjgl import TextureLoader
from com.mtgames.platformer.settings import Settings
from com.mtgames.platformer.debug import Command
from com.mtgames.platformer.gfx.gui import GUI

from com.mtgames.platformer.entities import LightSource

from java.lang import Math
from com.sun.javafx.geom import Vec3f

class Player(EntityInterface):
    def __init__(self):
        self.input = 0
        self.speed = 0
        self.maxHealth = 0
        self.jumpWaitMax = 0
        self.jumpSpeed = 0
        self.dashSpeed = 0
        self.dashWaitMax = 0
        self.staggerLength = 0

        self.xa = 0
        self.ya = 0
        self.xaDash = 0
        self.jumpWait = 0
        self.dashWait = 0
        self.dashTime = 0
        self.staggerTime = 0

        self.canJump = False
        self.canDash = True
        self.isStaggered = False
        self.isDashing = False
        self.dashDeplete = False

        self.textureID = TextureLoader.loadTextureArray("/assets/graphics/entities/player", 8)

        self.lightSource = 0


    def init(self, entity):
        entity.collide = True
        entity.persistent = True

        self.input = entity.getProperties().getInput()
        self.speed = entity.getProperties().getSpeed()
        self.maxHealth = entity.getProperties().getMaxHealth()
        self.jumpWaitMax = entity.getProperties().getJumpWait()
        self.jumpSpeed = entity.getProperties().getJumpSpeed()
        self.dashSpeed = entity.getProperties().getDashSpeed()
        self.dashWaitMax = entity.getProperties().getDashWait()
        self.staggerLength = entity.getProperties().getStaggerLength()

        entity.life = self.maxHealth

        self.lightSource = LightSource(entity.x, entity.y, entity.getProperties())
        entity.getProperties().getLevel().addLightSource(self.lightSource)


    def tick(self, entity):
        self.isStaggered = self.staggerTime != 0

        if not self.isStaggered:
            if entity.hasCollided(0, 1) and self.jumpWait > self.jumpWaitMax:
                self.canJump = True
                self.jumpWait = 0

            if entity.hasCollided(0, 1) and not self.canJump:
                self.jumpWait += 1

            if not entity.hasCollided(0, 1):
                self.canJump = False

            if self.input.isPressed(Settings.KEY_JUMP) and self.canJump and entity.isAlive():
                self.ya = -self.jumpSpeed
                self.canJump = False
                entity.animationFrame = 0

            if self.input.isPressed(Settings.KEY_DASH) and self.canDash and entity.isAlive():
                self.xaDash = self.dashSpeed
                self.canDash = False
                self.dashDeplete = True
                entity.animationFrame = 0

            if self.input.isPressed(Settings.KEY_LEFT) and entity.isAlive() and not self.isDashing:
                self.xa -= self.speed

            if self.input.isPressed(Settings.KEY_RIGHT) and entity.isAlive() and not self.isDashing:
                self.xa += self.speed

            if self.input.isPressed(Settings.KEY_TORCH) and entity.isAlive():
                Command.queue("light torch %d %d {\"colour\":%d}" % (entity.x, entity.y, int((Math.random() * 0xffffff))))
                self.input.unset(Settings.KEY_TORCH)

        else:
            self.staggerTime -=1
            self.ya = 0

        self.isDashing = self.xaDash != 0

        if self.xa > 1:
            self.xa -= 2
        elif self.xa < -1:
            self.xa += 2
        elif self.xa == 1 or self.xa == -1:
            self.xa = 0

        if self.xa > self.speed and not self.isDashing:
            self.xa = self.speed
        elif self.xa < -self.speed and not self.isDashing:
            self.xa = -self.speed

        entity.move(self.xa, self.ya)
        self.dash(entity)
        self.ya = entity.gravity(self.ya)

        if self.isDashing:
            for _ in range(10):
                Command.queue("light dashParticle %d %d" % (entity.x, entity.y))

        if entity.hasCollidedEntity("baseEnemy"):
            entity.life -= 1
            self.staggerTime = self.staggerLength


        self.lightSource.move(entity.x, entity.y)


    def render(self, entity, screen):
        xtile = 0
        flipx = False

        if entity.movingDir == 0:
            flipx = True

        if entity.isJumping and not self.isDashing and not self.isStaggered:
            xtile = 4
        elif self.isDashing:
            xtile = 6
            entity.animationFrame = 0
        elif self.isStaggered:
            xtile = 7
            entity.animationFrame = 0

        xtile += entity.animationFrame

        screen.renderEntity(entity.x, entity.y, self.textureID[xtile], 16, flipx)

        GUI.add(lambda : GUI.progressBar(80, 13, 16, 150, float(entity.life) / self.maxHealth, Vec3f(0.4, 0.1, 0.1)))
        dashratio = float(self.dashWait)/self.dashWaitMax
        if dashratio == 1:
            colour = Vec3f(0.1, 0.1, 0.5)
        else:
            colour = Vec3f(0.1, 0.1, 0.4)

        GUI.add(lambda : GUI.progressBar(screen.width-80, 12, 16, 150, dashratio, colour))

        # TODO: Temp death code
        if not entity.isAlive():
            GUI.add(lambda : GUI.textBox("You died!", ""))


    def dash(self, entity):
        if self.dashWait >= self.dashWaitMax:
            self.dashWait = self.dashWaitMax
        else:
            self.dashWait += 1

        self.canDash = self.dashWait == self.dashWaitMax and not entity.hasCollided(0, 1)

        if self.dashDeplete:
            self.dashWait -= 5

        if self.dashWait <= 0:
            self.dashDeplete = False

        if self.xaDash == 0:
            return

        if entity.hasCollided(-1, 0) or entity.hasCollided(1, 0):
            self.staggerTime = self.staggerLength
            self.dashDeplete = False
            self.xaDash = 0
            entity.life -= 10

        self.ya = 0
        entity.gravityWait = 0

        if entity.movingDir == 0:
            self.xa = -self.xaDash * self.speed
        else:
            self.xa = self.xaDash * self.speed

        if self.dashTime > 2:
            self.xaDash -= 1
            self.dashTime = 0

        self.dashTime += 1
